### DISCUSSION TITLE
云存储支持oss/cos等

### DISCUSSION THREAD

**Discussion by skyoct** [labring/laf#196] [Generals]
静态资源一般都比较大占用服务器大量带宽，而服务器带宽又是比较贵的，相对来说使用oss/cos来存储静态文件可能比较划算。
抽象出云存储需要使用到接口，比如创建桶、上传文件、下载文件、自定义域名等。
再根据不同接口实现oss/cos 这些对应的功能。

**Comment by maslow** [comment]
我觉得可以考虑，特别是在 laf admin 开发好之后，可以在 laf admin 中配置云存储的后端。

可能需要考虑的几个注意点：
1. 这种抽象的难度以及 局限性 会有多大，厂商的对接增加了复杂度，各家 oss 虽然类 s3 接口 但又不是遵循 s3 标准
2. 云厂商「创建和管理 bucket」的接口并非 oss 操作接口，而属于 platform open api 范畴，是受限的；
3. 云厂商 oss 服务对一个账户创建的 bucket 、域名数量 有严格的限制
4. laf app 当前依然可以使用 云厂商的 oss，就像传统开发方式那样，直接在云函数调用即可，我们考虑实现的应该是 将 oss/cos 集成到 laf 控制台？

**Reply by maslow** [reply]
我对这个抽象还没有更多想法，比较模糊，或者 你再提供 更进一步的想法？

**Reply by skyoct** [reply]
1、2. 大概就是 我们把laf可能用到接口都定义好，
然后minio/cos/oss等 客户端根据接口去实现对应功能 
接口大概包含创建桶/删除桶/创建目录/删除目录/上传文件/删除文件/自定义域名
这块可以复用的接口可以选择默认实现，和平台强关联的接口就需要单独实现
3. 对于 bucket 、域名数量有限制这块 我大概看了下，cos和oss貌似创建桶在100个，自定义域名在20个 基本上是可以满足小团队/用户需求，当有更大量需求客户应该是可以给阿里云/腾讯云提工单要求提升。或者有特殊需求的用户就可以选择minio作为云存储。
4. 现在是可以直接调用oss/cos, 但是laf没办法直接管理，还需要登录阿里云/腾讯云去管理，没办法做到“一站式”开发。

**Reply by skyoct** [reply]
限额问题：用户配置云存储的时候我们就可以注明oss/cos可能存在限额，如果有比较高的需求，尽量使用minio

**Reply by maslow** [reply]
OK，是否可以得出结论，这样做的主要原因：
 1. 扩展场景方案：支持更多云存储方案，以满足更灵活场景需求
 2. 一站式开发体验：需要 laf 可直接管理云存储，像管理内置的 minio 那样

q: 对于（1）能不能举/假设一个更具体点的例子/场景?

**Reply by skyoct** [reply]
- 创建bucket定义为function putBucket(bucketName)，我们根据s3规范实现一份，cos和oss支持s3的创建bucket就可以直接用，如果自己实现一个存储怎么办，其中并不支持s3怎么办（比如基于ceph自建存储），那么就必须自己实现这个接口。

- 绑定自定义域名定义为function setBucketCustomDomain(buketName, domain), 这个可能属于平台特有的，oss nodejs SDK没这个方法，但是有这个接口，oss需要实现这个方法，调用oss接口完成绑定自定义域名。cos nodejs SDK有绑定自定义域名的方法：putBucketDomain，那么cos实现setBucketCustomDomain方法内部直接调用putBucketDomain方法。其他类似。

并不需要抽象到很高程度，我们目前对于存储应用场景是有限的....

**Comment by virgoone** [comment]
阿里云OSS是支持S3接口的，通过s3的SDK可以一键无缝上传至阿里oss，我在自己的云函数中实现过

**Reply by virgoone** [reply]
同：七牛云存储也兼容S3的API的

**Reply by skyoct** [reply]
不仅要考虑创建桶、上传、下载、删除，还要考虑静态网站托管和自定义域名。
其中七牛云应该是不支持API自定义域名的

**Comment by youmengme** [comment]
这个还有后续吗 期待ing~

**Reply by maslow** [reply]
有的，其中有相关产品的官方团队已经联系了 laf， 表示想给 laf 贡献支持其oss 产品的实现

**Comment by GiantappMan** [comment]
蹲个文档

**Comment by m430** [comment]
2023年这个出来了吗

**Reply by skyoct** [reply]
目前还不支持
